### PR TITLE
feat: advanced settings page

### PR DESCRIPTION
**Pull request by sshanzel**
DD-245 #done

Advanced Settings page:
- Created a new property in `feedSettings` which will hold the actual list of advanced settings and the user's preferences itself.
- New component that might be used for future switches namely `FilterSwitch`. The component requires a `title` and `description`.

Enabled = true:
Screen Shot 2021-10-29 at 4 47 14 PM

Enabled = false:
Screen Shot 2021-10-29 at 4 47 44 PM

Notice the first few articles were tagged as part of "Tech magazines" in my local DB and it went away after the change.

### CODE REVIEW

**Review by idoshamun — CHANGES_REQUESTED** [review]
Trying it out here as a non-registered user, none of the toggles are turned on although all of them are enabled by default
https://daily-webapp-7a9kb6cdt-dailydotdev.vercel.app/

And also check my comments

File: packages/shared/src/components/filters/AdvancedSettings.tsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    });
+  };
+
+  useEffect(() => {
+    if (!feedSettings?.advancedSettings) {
+      return;
+    }
+
+    const map = {};
+    feedSettings.advancedSettings.forEach(
```
[the remark this answers is not in the document]
Comment (sshanzel) [reply packages/shared/src/components/filters/AdvancedSettings.tsx:33]: I have to agree. This is the best case to utilize those 2 functions.

File: packages/shared/src/components/filters/FilterSwitch.tsx [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import React, { ReactElement } from 'react';
+import { Switch, SwitchProps } from '../fields/Switch';
+
+interface FilterSwitchProps extends SwitchProps {
+  title: string;
+  description: string;
+}
+
+export function FilterSwitch({
+  title,
```
[the remark this answers is not in the document]
Comment (sshanzel) [reply packages/shared/src/components/filters/FilterSwitch.tsx:10]: Totally right. I shall make these changes.

**Comment by sshanzel** [comment]
Hi @idoshamun thank you for the awesome review! I have now added a test for checking the merged values and for unregistered ones 🚢

**Comment by sshanzel** [comment]
Hi @idoshamun thank you for the awesome review! I have now added a test for checking the merged values and for unregistered ones 🚢

File: packages/shared/src/components/filters/AdvancedSettings.spec.tsx [hunk 3]
```diff
@@ -27,14 +27,16 @@ import { getFeedSettingsQueryKey } from '../../hooks/useMutateFilters';
 import { waitForNock } from '../../../__tests__/helpers/utilities';
 
 const showLogin = jest.fn();
+// eslint-disable-next-line prefer-const, no-undef-init
```
Comment (idoshamun) [inline packages/shared/src/components/filters/AdvancedSettings.spec.tsx:30]: Why is this necessary? You don't actually need to initialize it with `undefined` because this is the default value.
And also eslint should not complain about `let` if you actually use it.

File: packages/shared/src/components/filters/AdvancedSettings.tsx [hunk 4]
```diff
@@ -24,31 +32,21 @@ function AdvancedSettingsFilter(): ReactElement {
...
-          inputId={`${title}-${id}`}
-        />
-      ))}
+      {advancedSettings?.map(
+        ({ id, title, description, defaultEnabledState }) => (
+          <FilterSwitch
+            key={id}
+            label={title}
+            checked={settings[id] ?? defaultEnabledState}
+            name={ADVANCED_SETTINGS_KEY}
```
Comment (idoshamun) [inline packages/shared/src/components/filters/AdvancedSettings.tsx:43]: You can't use the same input name for all the components. You can just use the same value like `inputId`

**Review by idoshamun — CHANGES_REQUESTED** [review]
Besides my inline comments, I also think we need another test case for when the user is not logged in

File: packages/shared/src/components/filters/AdvancedSettings.spec.tsx (hunk 3 above)
Comment (sshanzel) [reply packages/shared/src/components/filters/AdvancedSettings.spec.tsx:30]: Ah, I see, that could be why it was throwing an error.

**Comment by sshanzel** [comment]
Hi @idoshamun I have included a test case for when the user is not logged in by not setting a value to the `loggedUser`. Then I also included an expectation for the switch to be `checked`
Though I have changed the description for the mentioned test case to be more explicit.

https://github.com/dailydotdev/apps/blob/f26bf76f7168541f401594d9eff4b84af2d6db25/packages/shared/src/components/filters/AdvancedSettings.spec.tsx#L104

**Review by idoshamun — APPROVED** [review]
Woohoo! Probably I missed the test case before. Sorry for that.
Let's merge this PR! 💪

**Comment by sshanzel** [comment]
No worries at all! It was my mistake not being more explicit especially the case above it looks very identical which is for registered user :P 

Ship awayyy! 🚢
